### PR TITLE
[LDAP] Allow adding and removing values to/from multi-valued attributes

### PR DESCRIPTION
**Pull request by jean-gui**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

`EntryManagerInterface::update(Entry $entry)` is extremely slow in some specific cases such as adding or removing members to or from huge groupOfNames if you also enable the memberOf overlay in OpenLDAP. Disabling memberOf does make things a lot better, but it is still slow compared to inserting/removing only the elements you want.

This PR adds two methods to Symfony\Component\Ldap\Adapter\ExtLdap\EntryManager taking advantage of ldap_mod_add and ldap_mod_del.

I thought about using them directly in the update method, but since you need to know what values to add and remove, it would be necessary to retrieve the old values from LDAP.

I'm also unsure whether these two methods should be in an interface. I think that adding them to EntryManagerInterface would break BC, but I could create another interface, similarly to RenameEntryInterface.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
ping @csarrazi

**Comment by csarrazi** [comment]
I think that we could improve this by introducing a new API for handling changes.

Instead of using separate methods for different things, we should rather use a Command pattern, and provide a set of commands which could be implemented without breaking the interface each time we need to implement something.

This way, we could have a runCommand() method in the EntryManager, with a list of operations:
* UpdateEntry
* AddAttributes
* RemoveAttributes
* RenameEntry

This would support the current set, and we could deprecate the other methods afterwards.

What do you think about this @nicolas-grekas ?

**Comment by csarrazi** [comment]
By the way, this means that the `add()`, `update()`, `remove()` and `rename()` methods should all be deprecated, in favor of the `runCommand()` (Or `executeCommand()`) method.

**Comment by nicolas-grekas** [comment]
ping @jean-gui

**Comment by jean-gui** [comment]
I was actually waiting for your response to @csarrazi's ping. I'm not sure I have the necessary skills or time to implement this new API though. I'll try to, but am not making any promises.

**Comment by csarrazi** [comment]
@jean-gui want me to take over that one?

**Comment by csarrazi** [comment]
I won't have much time before next week, though.

@symfony/deciders What do you think about the suggestion, btw?

**Comment by jean-gui** [comment]
@csarrazi If you can, sure!

**Comment by nicolas-grekas** [comment]
Let's see how it goes for 3.4 :)

**Comment by nicolas-grekas** [comment]
Status: needs work

**Comment by fabpot** [comment]
@csarrazi Do you think you will have time to work on this one soon? If not, anybody else?

**Comment by csarrazi** [comment]
Hi @fabpot. Unfortunately I won't have time to work on this anytime soon. I can spend some time reviewing PRs, but unfortunately I have little time to do anything else.

**Comment by nicolas-grekas** [comment]
Moving to 4.1. Rebase on master might be needed, where PHP 7.1 features can be used btw.

**Comment by fabpot** [comment]
Anyway willing to take over this one? If not, I'm going to close.

**Comment by jean-gui** [comment]
@fabpot If no one is willing to tackle the big refactoring suggested by @csarrazi, perhaps this PR could be considered for merging.

**Comment by csarrazi** [comment]
We can do that. Like I mentioned before, I no longer contribute to Symfony on a regular basis.

One thing though. I would change the method name to removeAttributeValues / addAttributeValues.

Second, this PR should be updated to take into account changes in coding standards for Symfony 4 (PHP > 7.1.3 is used, so scalar type hinting can be used now).

By the way, `$values` should be type hinted as an `array`.

**Comment by jean-gui** [comment]
I'll try to do that today.

File: src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php
Comment (fabpot) [inline src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php]: The 3 `@param` entries can be removed (we don't add those if they don't add something that is not part of the PHP method signature already).

File: src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php
Comment (fabpot) [inline src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php]: LDAP

File: src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php
Comment (fabpot) [inline src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php]: should be on one line.

Comment (fabpot) [reply src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php]: missing dot at the end of the error message.

File: src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php
Comment (fabpot) [inline src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php]: LDAP

File: src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php
Comment (fabpot) [inline src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php]: same as above

File: src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php
Comment (fabpot) [inline src/Symfony/Component/Ldap/Adapter/ExtLdap/EntryManager.php]: should also be on one line, and dot missing at the end of the error message.

**Review by fabpot — APPROVED** [review]
some small comments (mainly coding standard).

**Comment by jean-gui** [comment]
Done. Should I add those new methods to EntryManagerInterface? Would that be a BC break?

**Comment by fabpot** [comment]
Adding them to the interface would be a BC break.

**Comment by fabpot** [comment]
One last thing before merging: can you add a note about these 2 new methods in the composer CHANGELOG file?

**Comment by jean-gui** [comment]
Done (if src/Symfony/Component/Ldap/CHANGELOG.md is the file you meant).

I should probably submit a PR for the doc as well.

**Review by fabpot — APPROVED** [review]
A PR on the docs would be the cherry on the cake :)

**Comment by fabpot** [comment]
Thank you @jean-gui.

**Pull request merged**
